x-pack/plugins/security/server/authentication/index.ts

```diff
@@ -105,28 +105,30 @@ export async function setupAuthentication({
       // authentication (username and password) or arbitrary external page managed by 3rd party
       // Identity Provider for SSO authentication mechanisms. Authentication provider is the one who
       // decides what location user should be redirected to.
-      return t.redirected(authenticationResult.redirectURL!);
+      return response.redirected(undefined, {
+        headers: {
+          location: authenticationResult.redirectURL!,
+        },
+      });
     }
 
-    let error;
     if (authenticationResult.failed()) {
       authLogger.info(`Authentication attempt failed: ${authenticationResult.error!.message}`);
-      error = wrapError(authenticationResult.error);
-
-      const authResponseHeaders = authenticationResult.authResponseHeaders;
-      for (const [headerName, headerValue] of Object.entries(authResponseHeaders || {})) {
-        if (error.output.headers[headerName] !== undefined) {
-          authLogger.warn(`Server rewrites a error response header [${headerName}].`);
-        }
-        // Hapi typings don't support headers that are `string[]`.
-        error.output.headers[headerName] = headerValue as any;
-      }
+      const error = authenticationResult.error;
+      return getErrorStatusCode(error) === 400
+        ? response.badRequest(error, {
+            headers: authenticationResult.authResponseHeaders,
+          })
+        : response.unauthorized(error, {
+            headers: authenticationResult.authResponseHeaders,
+          });
     } else {
       authLogger.info('Could not handle authentication attempt');
-      error = Boom.unauthorized();
+      const error = Boom.unauthorized();
```

Comment:
__nit:__ there is no need in a dedicated `error` variable, just pass `Boom.unauthorized()` directly to `response.unauthorized`.